plugins/gnupgplugin/gpgme_p.h

```diff
@@ -0,0 +1,277 @@
+ * Contact: Damien Caliste <dcaliste@free.fr>
+ * All rights reserved.
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#ifndef _FILE_OFFSET_BITS
+#define _FILE_OFFSET_BITS 64
+#endif
+#include <gpgme.h>
+
+#include <Crypto/key.h>
+#include <QString>
+#include <QDebug>
+#include <QDateTime>
+
+struct GPGmeContext {
+    gpgme_ctx_t ctx;
+    gpgme_error_t err;
+    GPGmeContext(gpgme_protocol_t protocol, const QString &home = QString())
+        : ctx(0), err(0)
+    {
+        err = gpgme_engine_check_version(protocol);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            gpgme_engine_info_t info;
+            gpgme_error_t err2;
+            err2 = gpgme_get_engine_info(&info);
+            if (!err2) {
+                while (info) {
+                    if (info->protocol == protocol) {
+                        qWarning() << "protocol:" << gpgme_get_protocol_name(info->protocol);
+                        if (info->file_name && !info->version)
+                            qWarning() << "engine" << info->file_name
+                                       << "not installed properly";
+                        else if (info->file_name && info->version && info->req_version)
+                            qWarning() << "engine" << info->file_name
+                                       << "version" << info->version
+                                       << "installed, but at least version"
+                                       << info->req_version << "required";
+                        else
+                            qWarning() << "unknow issue";
+                    }
+                    info = info->next;
+                }
+            } else {
+                qWarning() << "cannot get engine info:" << gpgme_strerror(err2);
+            }
+            return;
+        }
+
+        err = gpgme_new(&ctx);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            ctx = 0;
+            return;
+        }
+        err = gpgme_set_protocol(ctx, protocol);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            gpgme_release(ctx);
+            ctx = 0;
+            return;
+        }
+        if (home.isEmpty())
+            return;
+        gpgme_engine_info_t info = gpgme_ctx_get_engine_info(ctx);
+        while (info && info->protocol != protocol)
+            info = info->next;
+        if (!info) {
+            qWarning() << "cannot change home.";
+            return;
+        }
+        err = gpgme_ctx_set_engine_info(ctx, protocol, info->file_name,
+                                        home.toUtf8().constData());
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            gpgme_release(ctx);
+            ctx = 0;
+            return;
+        }
+    }
+    ~GPGmeContext()
+    {
+        if (ctx)
+            gpgme_release(ctx);
+    }
+    bool operator!()
+    {
+        return (ctx == 0);
+    }
+    operator gpgme_ctx_t() const
+    {
+        return ctx;
+    }
+    QString error()
```

Comment:
can this be const or is either gpgme_error_code(err) or gpgme_strerror(err) nonconst?